### ISSUE TITLE
Support routing for SPA web apps

### ISSUE THREAD

**Issue by jpflueger** [fermyon/spin-fileserver#26]
I'm working on a Vue.js single-page web app that will be hosted by the static fileserver and I'd like to suggest an improvement. My Vue.js app uses vue-router to provide routing inside my application and currently I can only use the "hash mode" routing where the URL's hash is used for routing. This is not ideal for SEO and deep linking inside my application. For instance, the basic vue template has routes for:

```
/      -> index.html -> HomePage.vue
/about -> ...        -> AboutPage.vue
```

But there is not a file at the `/about` path so it returns a 404 like it should. For my use case (and this is likely to cover all SPA web apps), it would be awesome if I could configure the fileserver to return the content for index.html with headers to use the browser cache. This would allow the vue-router component to route when the page is re-loaded and I can manually handle any 404 in the UI with a catch all route. [See the vue-router docs for more info](https://router.vuejs.org/guide/essentials/history-mode.html). 

## Option 1: URL Rewrites
Looking at other platforms like Netlify and Vercel, it seems like they both support URL re-writes at the file server level. Something like this would allow me to specify the priority in which paths are checked and also create a "fallback" or "catch all" route. Some samples for comparison:

```
# nginx
location / {
  try_files $uri $uri/ /index.html;
}

# netlify
/* /index.html 200

# vercel
{
  "rewrites": [{ "source": "/:path*", "destination": "/index.html" }]
}
```

I'm not sure what the best way to express this is for the spin.toml but it seems do-able. This seems like the most complete solution but it also comes with an additional level of complexity. 

## Option 2: Fallback file
This is the simplest option I can think of that would catch a "file not found" error and instead of returning a 404 it would return a file at some configurable path. Obviously this would have to be opt-in as you don't want to break functionality for existing users but supplying a `fallback_route` config item and checking for its existence should suffice. 

I have tried to solve this with the redirect module but I don't want the window's location to change so I think it needs to be done in the fileserver component unless I'm missing something else, open to any other options or ideas you may have. Thanks!

**Comment by jpflueger** [comment]
I've worked up the two options that I discussed above. I'm pretty new to Rust so it may need some careful review if we like either of the paths forward.

[Option 1: URL Rewrites](https://github.com/fermyon/spin-fileserver/compare/main...jpflueger:spin-fileserver:feature/url-rewrites)
[Option 2: Fallback file](https://github.com/fermyon/spin-fileserver/compare/main...jpflueger:spin-fileserver:feature/fallback-path)

**Comment by itowlson** [comment]
I like the URL rewrites approach but, as you note, it would be better to do this in the `spin.toml`, as environment variables won't travel with the application.  And I think the only place we can fit them in is in the component `config` section which, as we have discussed, is limited to string values.  There are three ways we could tackle this:

1. Enhance Spin to allow non-string config values.  This is the "right" solution, but has far-reaching implications - it would require design effort and would be a breaking change for applications.
2. Kludge the rewrites into multiple string values (`rewrite1`, `rewrite2`, etc.)
3. Have one config setting that, if present, points to a rewrite file supplied as an asset.  The nominated file would not be served but would contain rewrite rules in whatever format you felt best.

The trouble with all these is that, because the fileserver is stateless, you'd have to go through parsing all the rules every time.  I noticed your comment about capturing it with Wizer, but at the moment that would require users to take a local copy and remember to wizerise it every time they change the rules.  It would lose the convenience of being able to reference the server as a standard component.

A fallback file avoids all these problems for sure - you only need one config entry, and you only need to consult it if serving fails.  So if that meets your needs then that's they way to go.  But from your description of Vue routing I wasn't sure if it would?

(Another longer-term option could be to implement URL rewriting into the Spin HTTP infrastructure itself, as `[component.trigger]` settings.  That would allow the parsed rewrites to be cached for the duration of the process, but the caching might not be viable on Cloud - I'm not sure.)

**Comment by jpflueger** [comment]
> Enhance Spin to allow non-string config values. This is the "right" solution, but has far-reaching implications - it would require design effort and would be a breaking change for applications.

Another future option might be to allow pointing to a key-value namespace (when key-value support is released) and use that for storing re-write rules. 

> The trouble with all these is that, because the fileserver is stateless, you'd have to go through parsing all the rules every time. I noticed your comment about capturing it with Wizer, but at the moment that would require users to take a local copy and remember to wizerise it every time they change the rules. It would lose the convenience of being able to reference the server as a standard component.

I think if we could provide a sample `spin.toml` with a `component.build` command to "wizerize" the module, that would be an acceptable solution to me. 

> A fallback file avoids all these problems for sure - you only need one config entry, and you only need to consult it if serving fails. So if that meets your needs then that's they way to go. But from your description of Vue routing I wasn't sure if it would?

I believe that would work and am testing out my implementation now. If that is an acceptable short-term solution that the project would be willing to adopt, I'd be happy to clean up the code and contribute a PR?

**Comment by rajatjindal** [comment]
(pls ignore if this is not relevant)

I ran into similar problem with my `deploy-to-fermyon` cloud implementation, and implemented a http-redirect wasm module https://github.com/fermyon/deploy-to-fermyon/blob/main/http-redirect/redirect.go

and used in `spin.toml` as

```toml

[[component]]
id = "redirect-index-html"
source = "http-redirect/redirect.wasm"
[component.trigger]
route = "/"
[component.build]
command = "cd http-redirect && tinygo build -wasm-abi=generic -target=wasi -gc=leaking -no-debug -o redirect.wasm redirect.go"
```

and then in a catch all, we let the vue router take over. (I had a _.vue which would do internal routing, but i  can't seem to find it, will let u know if i find it), 

but i think that was the idea in a nut shell.

**Comment by jpflueger** [comment]
Thanks Rajat! The redirect didn't work for me because I wanted the browser's location to remain the same while still rendering the index.html page. 

I think that long-term it sounds like the url re-write method might be the way to go but not until either complex config values are supported (like arrays) or _potentially_ key-value storage is supported. For now I have a local copy of my changes in [this branch](https://github.com/fermyon/spin-fileserver/compare/main...jpflueger:spin-fileserver:feature/fallback-path) running and it seems to work just fine for my Vue.js app. I can link to the app if you are interested.

**Comment by jpflueger** [comment]
@SamiKouatli I think we fixed this issue in #27 and #30 but forgot to resolve this issue. It looks like we need to cut a new release but it should be resolved by v0.0.3.
